### PR TITLE
add mpiexec to mpi providers

### PR DESCRIPTION
**Pull request by davydden**

### CODE REVIEW

File: var/spack/repos/builtin/packages/dealii/package.py [hunk 1]
```diff
@@ -312,6 +312,7 @@ def cmake_args(self):
                 '-DCMAKE_C_COMPILER=%s' % spec['mpi'].mpicc,
                 '-DCMAKE_CXX_COMPILER=%s' % spec['mpi'].mpicxx,
                 '-DCMAKE_Fortran_COMPILER=%s' % spec['mpi'].mpifc,
+                '-DMPIEXEC_EXECUTABLE=%s' % spec['mpi'].mpiexec
```
Comment (davydden) [inline var/spack/repos/builtin/packages/dealii/package.py]: @tamiko could you please confirm this is the right way to provide `mpiexec` to deal.II build system? I could not find much info in our docs...

@jppelteret FYI.

**Comment by davydden** [comment]
@alalazo 

> mvapich2 has not being modified: is it intentional?

no, just overlooked.

> I think there could be potential issues with this in the case any of the mpi provider has support for Slurm.

I don't have much first hand experience for that, so I will let you guys to decide on what's the best course of action. 

The reason I need some solution like that is to make sure that in dealii we won't pick up an unrelated `mpiexec`, i.e. `openmpi` provided by Ubuntu. Everything will build, but when running tests (including downstream projects) you are screwed ;-)

**Comment by chuckatkins** [comment]
@alalazo the issue you raise is relevant beyond slurm and is pertinent to other scenarios as well, all to do with how the resource manager is set up at the given site.  Cray MPICH environments might be using ALPS (via PBSPro, Torque, or Cobalt) like on Theta at ALCF or at DoD HPCMP, in which case they'd use `aprun` instead, or if they're using Slurm like on Cori at NERSC then they'd use `srun`.  Similarly using Spectrum  MPI on Summit at OLCF with LSF will use `jsrun` but could use `mpiexec` directly on a different cluster.

I suppose for a "system" provided MPI like cray-mpich or spectrum-mpi then you could rely on the site's packages.py to supply mpiexec correctly, but is there some mechanism in spack to supply that information to a package at install time?  For example, if you built the openmpi spack package on summit then how would tell it that mpiexec needs to be `jsrun`?

**Comment by alalazo** [comment]
Closing as stale. Feel free to reopen if you disagree and want to push this forward.

**Pull request closed without merging**